src/items/functions.md

```diff
@@ -214,12 +214,43 @@ extern "C" fn new_i32() -> i32 { 0 }
 let fptr: extern "C" fn() -> i32 = new_i32;
 ```
 
-Functions with an ABI that differs from `"Rust"` do not support unwinding in the
-exact same way that Rust does. Therefore, unwinding past the end of functions
-with such ABIs causes the process to abort.
-
-> **Note**: The LLVM backend of the `rustc` implementation
-aborts the process by executing an illegal instruction.
+### Unwinding
+
+Most ABI strings come in two variants, one with an `-unwind` suffix and one without.
+The `Rust` ABI always permits unwinding, so there is no `Rust-unwind` ABI.
+
+The choice of ABI, together with the [panic mode][panic-modes], determines the
+behavior when unwinding out of a function.
+
+The table below indicates the behavior of an unwinding operation reaching each
+type of ABI boundary (function declaration or definition using the
+corresponding ABI string). Note that the Rust runtime is not affected by, and
+cannot have an effect on, any unwinding that occurs entirely within another
+language's runtime, that is, unwinds that are thrown and caught without
+reaching a Rust ABI boundary.
+
+The "unwinding" ABI category refers to `"Rust"` (the implicit ABI of Rust
+functions not marked `extern`), `"C-unwind"`, and any other ABI with `-unwind`
+in its name. The "non-unwinding" ABI category refers to all other ABI strings,
+including `"C"` and `"stdcall"`.
+
+Native unwinding is defined per-target. On targets that support throwing and
+catching C++ exceptions, it refers to the mechanism used to implement this
+feature. Some platforms implement a form of unwinding referred to as ["forced
+unwinding"][forced-unwinding]; `longjmp` on Windows and `pthread_exit` in
+`glibc` are implemented this way. Forced unwinding is explicitly excluded
+from the "Native unwind" column in the table.
+
+| panic runtime  | ABI           | `panic`-unwind                        | Native unwind (unforced) |
+| -------------- | ------------  | ------------------------------------- | -----------------------  |
+| `panic=unwind` | unwinding     | unwind                                | unwind                   |
+| `panic=unwind` | non-unwinding | abort                                 | [undefined behavior]     |
+| `panic=abort`  | unwinding     | `panic!` aborts (no unwinding occurs) | abort                    |
+| `panic=abort`  | non-unwinding | `panic!` aborts (no unwinding occurs) | [undefined behavior]     |
```

Comment:
```suggestion
| `panic=unwind` | non-unwinding | abort (no unwinding) | [undefined behavior]     |
| `panic=abort`  | unwinding     | abort (no unwinding) | abort                    |
| `panic=abort`  | non-unwinding | abort (no unwinding) | [undefined behavior]     |
```

Is there meant to be any semantic difference between the first of these rows and the latter two with respect to the third column?  If not, probably their contents should be identical for clarity.

Reply:
Yes, but my understanding has changed since I originally wrote this table, so it might need revision.

In the `panic=unwind` case, it's possible that some or all destructors *prior to hitting an `extern "C"` boundary* are invoked. That is, some unwinding occurs, and then, partway through, the process aborts.

Conversely, in the `panic=abort` case, no unwinding occurs *at all* when Rust code panics.

The thing I didn't understand originally is the issue about not being able to guarantee whether or not destructors do in fact run in the `panic=unwind` case. So that cell is really indicating that an abort happens, and some amount of unwinding may or may not occur before that happens.

Reply:
Yes, as you say, we should clarify this and probably footnote that cell in the table.  Perhaps that'd also be the place to mention the carve-out for 2-phase unwinding.  I.e., with `panic=unwind` and a non-unwinding `extern fn` ABI, then after landing https://github.com/rust-lang/rust/pull/129582, either we'll unwind to that boundary or will abort without unwinding.

Reply:
> In the panic=unwind case, it's possible that some or all destructors prior to hitting an extern "C" boundary are invoked. That is, some unwinding occurs, and then, partway through, the process aborts.

The intended semantics -- assuming https://github.com/rust-lang/rust/pull/129582 -- is that there are only two cases:
- we abort before unwinding starts
- we unwind all the way to the `extern "C"` boundary

So that's probably what the documentation should say?

Reply:
Ah, I didn't see that https://github.com/rust-lang/rust/pull/129582 had already been approved by the lang team.

Footnote added. (And other miscellaneous adjustments.)

Reply:
Marking this resolved since the question of whether this should be a footnote or something more prominent is in a separate thread.